File: src/sim/pv/losses.ts
import type { LossComponent, SystemLoss } from '../../types/energy'
import type { Fraction } from '../../types/units'

const loss = (component: LossComponent, fraction: number, label: string): SystemLoss => ({
  component,
  fraction: fraction as Fraction,
  label,
})

/**
 * The PVWatts v5 default loss stack. Combined multiplicatively these give the
 * published 14.08% total, which is what the unit test pins
 */
export const PVWATTS_DEFAULT_LOSSES: readonly SystemLoss[] = [
  loss('soiling', 0.02, 'Soiling: dust, pollen and bird droppings on the glass'),
  loss('shading', 0.03, 'Shading: near obstructions and row-to-row losses'),
  loss('snow', 0, 'Snow: modules covered by settled snow'),
  loss('mismatch', 0.02, 'Mismatch: manufacturing spread between series-connected modules'),
  loss('wiring', 0.02, 'Wiring: resistive losses in DC and AC conductors'),
  loss('connections', 0.005, 'Connections: resistive losses at connectors and terminals'),
  loss(
    'light-induced-degradation',
    0.015,
    'Light-induced degradation: first-hours loss in crystalline silicon',
  ),
  loss('nameplate', 0.01, 'Nameplate rating: measured power below the label value'),
  loss('age', 0, 'Age: cumulative degradation over the years in service'),
  loss('availability', 0.03, 'Availability: time out of service for maintenance and outages'),
]

export const overrideLoss = (
  losses: readonly SystemLoss[],
  component: LossComponent,
  fraction: number,
  label?: string,
): readonly SystemLoss[] =>
  losses.map((entry) =>
    entry.component === component
      ? { ...entry, fraction: fraction as Fraction, label: label ?? entry.label }
      : entry,
  )

export const lossDerateFactor = (losses: readonly SystemLoss[]): number =>
  losses.reduce((factor, entry) => factor * (1 - entry.fraction), 1)

export const combinedLossFraction = (losses: readonly SystemLoss[]): Fraction =>
  (1 - lossDerateFactor(losses)) as Fraction

File: crates/agv-sim/src/pv/losses.rs
//! The PVWatts v5 loss stack, ported from `src/sim/pv/losses.ts`.
//!
//! Labels stay on the TypeScript side. They are user-facing prose, they would be the only strings
//! crossing this crate's boundary, and `src/wasm.rs` explains why that boundary is f64 only. What
//! is shared is the arithmetic and the published fractions, which is what can disagree.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LossComponent {
    Soiling,
    Shading,
    Snow,
    Mismatch,
    Wiring,
    Connections,
    LightInducedDegradation,
    Nameplate,
    Age,
    Availability,
}

/// The PVWatts v5 defaults. Combined multiplicatively these give the published 14.08% total,
/// which is what the unit test pins.
pub const PVWATTS_DEFAULT_LOSSES: &[(LossComponent, f64)] = &[
    (LossComponent::Soiling, 0.02),
    (LossComponent::Shading, 0.03),
    (LossComponent::Snow, 0.0),
    (LossComponent::Mismatch, 0.02),
    (LossComponent::Wiring, 0.02),
    (LossComponent::Connections, 0.005),
    (LossComponent::LightInducedDegradation, 0.015),
    (LossComponent::Nameplate, 0.01),
    (LossComponent::Age, 0.0),
    (LossComponent::Availability, 0.03),
];

/// What survives the stack, as a multiplier.
pub fn loss_derate_factor(losses: &[(LossComponent, f64)]) -> f64 {
    losses.iter().fold(1.0, |factor, (_, f)| factor * (1.0 - f))
}

pub fn combined_loss_fraction(losses: &[(LossComponent, f64)]) -> f64 {
    1.0 - loss_derate_factor(losses)
}

/// Replace one component's fraction, leaving the order and the rest untouched.
pub fn override_loss(
    losses: &[(LossComponent, f64)],
    component: LossComponent,
    fraction: f64,
) -> Vec<(LossComponent, f64)> {
    losses
        .iter()
        .map(|(c, f)| {
            if *c == component {
                (*c, fraction)
            } else {
                (*c, *f)
            }
        })
        .collect()
}
